pkg/leech/pipeline_test.go

```diff
@@ -0,0 +1,228 @@
+	t.Parallel()
+
+	ctx := context.Background()
+
+	testPrivateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cases := []struct {
+		name             string
+		repoName         string
+		logPath          string
+		gcsPath          string
+		wantErr          string
+		tokenHandlerFunc http.HandlerFunc
+		ghHandlerFunc    http.HandlerFunc
+		writerFunc       func(context.Context, io.Reader, string) error
+	}{
+		{
+			name:     "success",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+		},
+		{
+			name:     "failed_access_token_generation",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			tokenHandlerFunc: func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+			},
+			wantErr: "error getting GitHub access token",
+		},
+		{
+			name:     "github_general_failure",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			ghHandlerFunc: func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+				fmt.Fprintf(w, "test bad request")
+			},
+			wantErr: `error response from GitHub - response body: "test bad request"`,
+		},
+		{
+			name:     "github_logs_not_found",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			ghHandlerFunc: func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusNotFound)
+				fmt.Fprintf(w, "not found")
+			},
+			wantErr: "GitHub logs expired",
+		},
+		{
+			name:     "github_logs_gone",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			ghHandlerFunc: func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusGone)
+				fmt.Fprintf(w, "gone")
+			},
+			wantErr: "GitHub logs expired",
+		},
+		{
+			name:     "object_write_bad_url",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "HOT GARBAGE",
+			wantErr:  "error copying logs to cloud storage: malformed gcs url: invalid uri: [HOT GARBAGE]",
+		},
+		{
+			name:     "object_write_failure",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			writerFunc: func(ctx context.Context, r io.Reader, s string) error {
+				return fmt.Errorf("write failed")
+			},
+			wantErr: "error copying logs to cloud storage: write failed",
+		},
+		{
+			name:     "read_write_match",
+			repoName: "test/repo",
+			logPath:  "test/repo/logs",
+			gcsPath:  "gs://test/repo/logs/artifacts.tar.gz",
+			ghHandlerFunc: func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(200)
+				fmt.Fprintf(w, "test-results")
+			},
+			writerFunc: func(ctx context.Context, r io.Reader, s string) error {
```

Comment:
It's better to compare results explicitly in test. Here I would just have `testObjectWriter` to always catch the written bytes and then in the test, do:

```go
if got, want := tc.testObjectWriter.gotArtifact, tc.wantArtifact; got != want {
  t.Errorf("artifacts written got=%v want=%v", got, want)
}
```

Reply:
👍